Add schema access rules documentation

hashhar opened the pull request:
Fixes #4665

hashhar commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -77,13 +77,15 @@ contents:
-* It contains the rules defining which catalog can be accessed by which user (see Catalog Rules below).
+* The catalog rules specifying which catalog can be accessed by which user (see Catalog Rules below).
+* The schema rules specifying which users have ownership of which schemas (see Schema Rules below).
+* The table rules specifying which privileges are granted to which users over which tables (see Table Rules below).
```
This is very "wordy". Any simpler suggestions?

mosabua replied:
Yes .. something like that .. make sure you have all anchors in place and such.

```
It contains rules that define which use has access to specific resources:

* :ref:`Catalogs <catalog-rules>` 
* :ref:`Schemas <schema-rules>`
```

hashhar replied:
Done.

hashhar commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+
+.. _table_rules:
+
+Table Rules
```
With the way the table rules are actually evaluated in the built-in SAC means that by if no table rules are provided all users are assumed to have all privileges (documented now in this PR).

In addition to this all tables in information_schema are always allowed regardless of the rules in the config file. Should this be documented?

mosabua replied:
information schema behavior should be documented

hashhar replied:
Documented within a note in the section (similar to catalog rules).

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -77,13 +77,15 @@ contents:
 * The system information rules specifying what users can access system management information (see System Information Rules below).
 
-This plugin currently supports catalog access, query, impersonation, principal, and system information
+This plugin currently supports catalog, schema, table access, query, impersonation, principal, and system information
```
This whole sentence can be removed after the above list is done and contains all of these.

hashhar replied:
Done.

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+users as owners of ``default`` schema and prevent user ``guest`` from ownership
+of any schema, you can use the following rules:
+
+.. literalinclude:: schema-access.json
```
Personally I would just inline the code snippet in a code block .. wdyt @electrum ?

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+Schema Rules
+------------
+
+These rules govern who is considered an owner of a schema. The user is granted
```
Is that correct? I dont think they provide ownership .. they just grant access to the schema right?

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+------------
+
+These rules govern who is considered an owner of a schema. The user is granted
+the privileges based on the first matching rule read from top to bottom. If no
```
The matching being done in order applies to all rules and should be explained in an earlier section.

hashhar replied:
Explained after the list of rules ends and have removed the mention of the ordering from each individual rule to avoid repetition.

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+Schema Rules
+------------
+
+These rules govern who is considered an owner of a schema. The user is granted
```
Schema rules allow you to grant ownership of a schema.

mosabua replied:
As a next sentence we should document what ownership of a schema means..

hashhar replied:
Changed to: 

> These rules allow you to grant ownership of a schema. Having ownership of an
> schema allows users to execute `DROP SCHEMA`, `ALTER SCHEMA` (both renaming
> and setting authorization), `SHOW CREATE SCHEMA`.

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+Table Rules
+-----------
+
+These rules govern the privileges granted to users on tables. The user is
```
Tables rules define privileges for table access.

hashhar replied:
Done.

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+* ``privileges`` (required): zero or more of ``SELECT``, ``INSERT``,
+  ``DELETE``, ``OWNERSHIP``, ``GRANT_SELECT``
+
+For example, to allow a user ``admin`` all privileges on all tables in all
```
Break this sentence up into a list of what happens... something like:

The following rules define the following table access:

* grant all privileges to the user ``admin`` on all tables in all schemas
...

hashhar replied:
Should I make this change to other rules too to be consistent?

mosabua replied:
Lets just do it here only for now

mosabua commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,6 +171,55 @@ following rules:
+privileges to all users on any table in ``default`` schema, you can use the
+following rules:
+
+.. literalinclude:: table-access.json
```
Probably also inline

electrum replied:
Agreed, these are short, and we're not referencing them in unit tests, so inline seems easier to read and maintain

hashhar commented:
@mosabua Addressed all comments except inlining of config files. I've kept it that way because it looks like all recent changes have used inlined files so there may be some reason. If not, do let me know and I'll inline others as well.

mosabua commented:
@electrum should decide about the inlining. Personally I would suggest we inline them since we do not use them for separate download or anything else from all I can tell.

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -75,18 +75,23 @@ contents:
-* The impersonation rules specify which user impersonations are allowed (see Impersonation Rules below).
-* The principal rules specifying what principals can identify as what users (see Principal Rules below).
-* The system information rules specifying what users can access system management information (see System Information Rules below).
+* :ref:`Catalogs <catalog_rules>`
```
We should avoid adding refs for links within the same document. They duplicate the built-in anchors for section titles, and the names are global within Sphinx, so care must be taken when naming. Instead, you can link to the built-in section anchors:
```
* `Catalogs <#catalog-rules>`__

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,14 +173,65 @@ following rules:
+
+.. note::
+
+    Users always have all privileges on all tables in the ``information_schema``.
```
This seems misleading, since metadata access like `information_schema` has its own security restrictions. Perhaps this should say
```
.. note:: These rules do not apply to ``information_schema``.

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,14 +173,65 @@ following rules:
+
+    Users always have all privileges on all tables in the ``information_schema``.
+
+The following rules define the following table access policy:
```
> The example below defines the following table access policy:

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -99,14 +104,13 @@ Presto restart. The refresh period is specified in the ``etc/access-control.prop
 Catalog Rules
 -------------
 
-These rules govern the catalogs particular users can access. The user is
```
It's fine to repeat the "top to bottom" information here. A user might visit this section directly, without having read the introduction at the top. It's one extra sentence, so repeating it doesn't hurt readability, and serves to reinforce the concept.

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,14 +173,65 @@ following rules:
+Schema Rules
+------------
+
+These rules allow you to grant ownership of a schema. Having ownership of an
```
Let's add the "top to bottom" information here to be consistent with other sections.

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,14 +173,65 @@ following rules:
+Table Rules
+-----------
+
+These rules define the privileges for table access for users. If no table rules
```
Add "top to bottom"

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -169,14 +173,65 @@ following rules:
 Query Rules
 -----------
 
-These rules control the ability of a user to execute, view, or kill a query. The user is
```
Same, keep "top to bottom"

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -207,8 +262,7 @@ authenticates using their credentials, and then submits a query as a different
 user.  When the user context is changed, Presto will verify the administrator
 is authorized to run queries as the target user.
 
-When these rules are present, the authorization is based on the first matching rule,
```
Same, keep "top to bottom"

electrum commented on presto-docs/src/main/sphinx/security/built-in-system-access-control.rst:
```diff
@@ -317,8 +371,7 @@ System Information Rules
 ------------------------
 
 These rules specify which users can access the system information management interface.
-The user is granted or denied access, based on the first matching rule read from top to
```
Same, keep "top to bottom"

hashhar commented:
@electrum Adjusted as requested and squashed commits into one.

electrum commented:
Thanks!
